### ISSUE TITLE
Metrics for train not consistent between CV and predict

### ISSUE THREAD

**Issue by anne-leroy-T** [pycaret/pycaret#862]
Hello,

Could you please run the following code :  

from pycaret.datasets import get_data
from pycaret.regression import *
data = get_data('boston')
s = setup(data, target = 'medv',silent=True)
my_model = create_model('xgboost')
plot_model(my_model)

train_data = data.iloc[get_config('X_train').index.to_list()]
r2_score(x["medv"], x["Label"])

You will notice the following :   

- For the create model CV results :

![image](https://user-images.githubusercontent.com/25195665/99674070-d64f6600-2a75-11eb-9c6d-1921b00f3a59.png)

- For the plot(my_model) :  

![image](https://user-images.githubusercontent.com/25195665/99674219-0eef3f80-2a76-11eb-8f89-ae7465564fe7.png)

- For the :  train_data = data.iloc[get_config('X_train').index.to_list()]   and r2_score(x["medv"], x["Label"])

![image](https://user-images.githubusercontent.com/25195665/99674399-46f68280-2a76-11eb-9fd0-1b42e54de511.png)

You will notice that the predict on train does not give the result of the CV, this is shown in the plot(my_model) and also in the predict on X_train.

(why do I need the predict on X_train would you ask ) in order to do the plot error for the train and not only for the test.
Anyway even should I not need this, the plot_model(my_model) gives incorrect R2 for the train.

Please note I think this is a tree algorithm issue only, as with 'lr' for instance, R2 train is more consistent to CV though not identical to mean.

For your convenience, I ran this on the Boston Dataset, but please note that on my personal project the CV R2 score was 0.30 and the predict on train and plot gave me 1.00 or near.

Maybe you should run this on another dirty dataset that gives you poor result in order to confirm ;-)

I hope I'm clear, if not, do not hesitate to ask further questions.

Thanks a lot for your precious help.

Anne

**Comment by anne-leroy-T** [comment]
Could you change the title to ... not consistent between CV and predict ON TRAIN ...    because of course, it's different with predict on test ....

**Comment by pycaret** [comment]
@anne-leroy-T I  see your confusion. 

The score that you see after the `create_model` is the averaged CV score. What you are trying to find out if you were to use your trained model on the training set what would be the score, which is what you see when you use `plot_model` with the `use_train_data=True`. See the example below that you can reproduce:

```
from pycaret.datasets import get_data
data = get_data('boston')

from pycaret.regression import *
s = setup(data, target = 'medv', silent = True)

xgboost = create_model('xgboost')
plot_model(xgboost, use_train_data = True) # it gives 1.0000 (see plot below)
X_original = data.iloc[get_config('X_train').index,]
predictions = predict_model(xgboost, data=X_original)

from pycaret.utils import check_metric
check_metric(X_original['medv'], predictions['Label'], metric = 'R2')

[OUTPUT] 1.0000
 
# repeat the same thing with LR

lr = create_model('lr')
plot_model(lr, use_train_data=True) #it gives 0.764 on train set
predictions_lr = predict_model(lr, data=X_original)
check_metric(X_original['medv'], predictions_lr['Label'], metric = 'R2')

[OUTPUT] 0.7639 
```

![image](https://user-images.githubusercontent.com/58118658/99677500-c587fb80-2a47-11eb-83d3-bb53ce201f3e.png)

![image](https://user-images.githubusercontent.com/58118658/99677526-cde03680-2a47-11eb-97e7-48cfa9c13e13.png)


Hope this helps.

**Comment by anne-leroy-T** [comment]
Hello,

I'm really sceptical about the fact that on my personal use case, the CV score is 0.30 and that the predict on train is 1.00 ... and this only for tree algorithm...

But if you say so ...

Maybe future will tell i'm true.

**Comment by pycaret** [comment]
@anne-leroy-T Decision Trees are always constructed in full which means that they are constructed until the leaf node that's why they are always fit. When you are checking CV, it trains the model on the N-1 fold and uses the Nth fold for prediction hence you get an error.

Hope this helps.

**Comment by anne-leroy-T** [comment]
That's why the tree depth parameter exists in order to limit overfitting and not go to the leaf nodes !  you should not overfit your train model
